### PR TITLE
Fix #5178: FluxC Editor media upload error/retry handling

### LINKED ISSUE #5178: FluxC - Media Integration - Editor media upload error handling/retrying
Retry error handling (media not in store, store media doesn't match remote, etc...)

### PR DESCRIPTION
**Pull request by aforcier**
Fixes #5178.

Not ready for merge yet.

Addresses a few issues with media uploads in the editor:

1. Enables retries for failed media uploads
2. Prevents completed and failed media from being re-enqueued for upload when more media are added to a post
3. Fixes a few issues with upload state comparisons due to case differences (`MediaUploadState.toString()` would be lowercase, while `MediaUploadState.name()` would be upper, and either method for `MediaModel.UploadState` would be uppercase) <-- I think later on we should drop `MediaUploadState` entirely, and beef it up a bit, maybe with a [`fromMedia` method](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/9e535369eb7dfef24c844e90443b3c634d3c4bde/fluxc/src/main/java/org/wordpress/android/fluxc/model/post/PostStatus.java#L63)

### CODE REVIEW

**Comment by aforcier** [comment]
Remaining TODO:

* [ ] Test everything for self-hosted sites
* [ ] Non-generic error handling: errors returned during upload are only given a type, and no message - but we display the message in that case, resulting in a `null` displayed in the failed image overlay

**Comment by aforcier** [comment]
Refactored error displays so we never end up showing `null`.

Currently, we'll display `error.message` if it's set (it won't be until https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/209 is addressed). For XML-RPC sites, this will be internationalized, but not for WP.com - we might want to focus on type in general.

**Comment by maxme** [comment]
I was able to test timeout errors and aborted request (using Charles throttling and breakpoints). I tried on wpcom and self hosted sites as well: ✅

**Comment by maxme** [comment]
> For XML-RPC sites, this will be internationalized, but not for WP.com 

I think there is a missing `&locale=` argument in the call to `https://public-api.wordpress.com/rest/v1.1/sites/$id/media/new/` - I opened this https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/324

> we might want to focus on type in general.

Yes when it's possible (error type from the response is sometimes very vague, in these cases we must use the error message).

**Comment by aforcier** [comment]
To close the loop on `error.message`:

We know that media errors are not localized over the WP.com REST API, whether or not we add the `locale` param: https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/324

We're about to have the `error.message` field populated, at least for WP.com sites (https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/325), which means if we rely on it we'll be showing an error in English to all users.

I refactored to stop using `error.message` in b7783b2 - but maybe it's best to revert this, rely on `error.message` when it's a `GENERIC_ERROR`, and hope we get this fixed server-side soon.

What do you think @maxme?

**Comment by aforcier** [comment]
Marking this as ready for review, though I may need to make a change depending on the decision re: https://github.com/wordpress-mobile/WordPress-Android/pull/5220#issuecomment-281068857.

**Comment by maxme** [comment]
> I refactored to stop using error.message in b7783b2 - but maybe it's best to revert this, rely on error.message when it's a GENERIC_ERROR, and hope we get this fixed server-side soon.
> 
> What do you think @maxme?

I think it's better to use `error.message` (if it's not empty) for `GENERIC_ERROR`, better to give the user some informations.

**Comment by aforcier** [comment]
> I think it's better to use error.message (if it's not empty) for GENERIC_ERROR, better to give the user some informations.

👍 done in 7dcb452

File: WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java [hunk 1]
```diff
@@ -273,8 +273,9 @@ private void cancelUpload() {
     }
 
     private void dispatchUploadAction(@NonNull final MediaModel media) {
-        AppLog.i(T.MEDIA, "Dispatching upload action: " + media.getTitle());
-        media.setUploadState(UploadState.UPLOADING.toString());
+        AppLog.i(T.MEDIA, "Dispatching upload action for media with id: " + media.getId() +
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java]: Nitpick but can we explicitly write "local id" here (and below), like in https://github.com/wordpress-mobile/WordPress-Android/blob/568616999496b328eacc29c921b64aa2aa97efad/WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java#L155-L155

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 2]
```diff
@@ -779,21 +779,32 @@ public void onUploadSuccess(MediaModel media) {
...
         String localMediaId = String.valueOf(media.getId());
-        if (error.type == MediaErrorType.GENERIC_ERROR) {
-            mEditorMediaUploadListener.onMediaUploadFailed(localMediaId, getString(R.string.tap_to_try_again));
-        } else {
-            mEditorMediaUploadListener.onMediaUploadFailed(localMediaId, error.message);
+
+        // Display custom error depending on error type
+        String errorMessage;
+        switch (error.type) {
+            case UNAUTHORIZED:
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: I just tried, it looks good, we should remove the "retry" button in this case, but we already have another issue for this.

**Comment by maxme** [comment]
Tested different configurations: throttled network, canceled requests, switching to airplane mode: I got expected behavior. 

Also reminded me of our auto-retry download after upload ;)
```
02-21 05:14:08.931 17055-17055/org.wordpress.android.beta D/WordPress-EDITOR: Image not loaded -- From line 1419 of file:///android_asset/ZSSRichTextEditor.js
02-21 05:14:08.931 17055-17055/org.wordpress.android.beta D/WordPress-EDITOR: Reloading image:4 - https://taliwuttprivatedotcom.files.wordpress.com/2017/02/wqqfytk.jpg?retry=4 -- From line 1410 of file:///android_asset/ZSSRichTextEditor.js
```

Only have the minor AppLog comment before we :shipit:

File: WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java (hunk 1 above)
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java]: Done in f78b9e6.

**Comment by aforcier** [comment]
> Also reminded me of our auto-retry download after upload ;)
```
02-21 05:14:08.931 17055-17055/org.wordpress.android.beta D/WordPress-EDITOR: Image not loaded -- From line 1419 of file:///android_asset/ZSSRichTextEditor.js
02-21 05:14:08.931 17055-17055/org.wordpress.android.beta D/WordPress-EDITOR: Reloading image:4 - https://taliwuttprivatedotcom.files.wordpress.com/2017/02/wqqfytk.jpg?retry=4 -- From line 1410 of file:///android_asset/ZSSRichTextEditor.js
```

hah, forgot about that one 😬

**Comment by maxme** [comment]
:shipit:

**Pull request merged**